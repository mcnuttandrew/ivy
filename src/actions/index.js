/* global chrome*/
const SECOND = 1000;
const MINUTE = SECOND * 60;
const HOUR = MINUTE * 60;
const DAY = HOUR * 24;

const buildEasyAction = type => payload => dispatch =>
  dispatch({type, payload});
export const clearSelection = buildEasyAction('clear-selection');

// example action
// export const setPageId = payload => dispatch => {
//   dispatch({type: 'set-page-id', payload});
//   getTreeForId(payload)
//     .then(result => dispatch({
//       type: 'get-tree-from-cache',
//       payload: {
//         data: result,
//         pageId: payload
//       }
//     }));
// };
