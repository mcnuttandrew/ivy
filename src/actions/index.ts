import {Dispatch} from 'redux';

import VegaDataPreAlias from 'vega-datasets';
const VegaData: {[key: string]: any} = VegaDataPreAlias;

import {Analyzer} from 'type-analyzer';
const {computeColMeta} = Analyzer;

// const SECOND = 1000;
// const MINUTE = SECOND * 60;
// const HOUR = MINUTE * 60;
// const DAY = HOUR * 24;

export interface GenericAction {
  (payload?: any): (dispatch: Dispatch) => void;
}
interface GenericActionCreator {
  (type: string): GenericAction;
}

const buildEasyAction: GenericActionCreator = type => payload => dispatch =>
  dispatch({type, payload});
// export const loadDataFromPredefinedDatasets = buildEasyAction(
//   'load-data-from-predefined',
// );

export const loadDataFromPredefinedDatasets: GenericAction = fileName => dispatch => {
  fetch(VegaData[fileName].url)
    .then(d => d.json())
    .then(d => {
      dispatch({
        type: 'recieve-data-from-predefined',
        payload: d,
      });

      dispatch({
        type: 'recieve-type-inferences',
        payload: computeColMeta(d),
      });
    });
};

export const changeSelectedFile: GenericAction = fileName => dispatch => {
  dispatch({
    type: 'change-selected-file',
    payload: fileName,
  });
  loadDataFromPredefinedDatasets(fileName)(dispatch);
};

// example action
// export const setPageId = payload => dispatch => {
//   dispatch({type: 'set-page-id', payload});
//   getTreeForId(payload)
//     .then(result => dispatch({
//       type: 'get-tree-from-cache',
//       payload: {
//         data: result,
//         pageId: payload
//       }
//     }));
// };
