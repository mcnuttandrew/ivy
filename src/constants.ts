// i know this file is bad and crazy i feel bad
export interface OptionValue {
  display: string;
  value: string;
}
export interface EncodingOption {
  optionType: string;
  options: OptionValue[];
  optionSetter: (spec: any, selectedOption: string) => any;
  optionGetter: (spec: any) => string;
  optionDefault: string;
  predicate: (spec: any) => boolean;
}

import {VegaMark} from './types';
export const marks: VegaMark[] = [
  'arc',
  'area',
  // 'image',
  // 'group',
  // 'line',
  // 'path',
  'rect',
  // 'rule',
  // 'shape',
  'symbol',
  'text',
  // 'trail',
];

// check if using the nested spec mode, if so, re-route into that subspec
const maybePrefixWithSpec = (spec: any, route: string[]): string[] =>
  Boolean(spec.getIn(['spec'])) ? ['spec', ...route] : route;

const justCountAgg = [
  {display: 'none', value: undefined},
  {display: 'count', value: 'count'},
];
const spatialAggs = [
  ...justCountAgg,
  {display: 'min', value: 'min'},
  {display: 'max', value: 'max'},
  {display: 'sum', value: 'sum'},
  {display: 'mean', value: 'mean'},
  {display: 'median', value: 'median'},
  {display: 'mode', value: 'mode'},
];
const toOption = (x: string): OptionValue => ({display: x, value: x});
const binningOptions = [
  'none',
  'count',
  'min',
  'mean',
  'sum',
  'bin', // TODO bin is handled incorrectly
  'max',
  'median',
].map(toOption);

const buildSpatialOptions = (
  dimension: string,
  options: OptionValue[],
): EncodingOption => ({
  optionType: 'Agg.',
  options,
  optionSetter: (spec, selectedOption) => {
    const route = maybePrefixWithSpec(spec, ['encoding', dimension]);
    const newSpec = spec.setIn([...route, 'aggregate'], selectedOption);
    const addType = (x: any) => x.setIn([...route, 'type'], 'quantitative');
    const hasField = newSpec.getIn([...route, 'field']);
    return hasField ? newSpec : addType(newSpec);
  },
  optionGetter: spec => {
    const route = maybePrefixWithSpec(spec, [
      'encoding',
      dimension,
      'aggregate',
    ]);
    return spec.getIn(route);
  },
  optionDefault: 'none',
  predicate: () => true,
});

const typeRoute = (dim: string) => ['encoding', dim, 'scale', 'type'];
const buildScaleOption = (dim: string): EncodingOption => ({
  optionType: 'Scale type',
  options: ['linear', 'log'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(maybePrefixWithSpec(spec, typeRoute(dim)), option),
  optionGetter: spec =>
    spec.getIn(maybePrefixWithSpec(spec, typeRoute(dim))) || 'linear',
  optionDefault: 'linear',
  predicate: spec =>
    Boolean(spec.getIn(maybePrefixWithSpec(spec, ['encoding', dim, 'field']))),
});

const buildCoerceRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'type']);
const buildTypeCoercion = (dim: string): EncodingOption => ({
  optionType: 'Data type',
  options: ['nominal', 'ordinal', 'quantitative', 'temporal'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(buildCoerceRoute(spec, dim), option),
  optionGetter: spec => spec.getIn(buildCoerceRoute(spec, dim)),
  optionDefault: undefined,
  predicate: () => true,
});

// take in an encoding option and create a predicate that rejects that option
// if that field is not present
type predicateInject = (dim: string, option: EncodingOption) => EncodingOption;
const buildPredRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'field']);
const injectFieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !!spec.getIn(buildPredRoute(spec, dim)),
});
const injectNofieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !spec.getIn(buildPredRoute(spec, dim)),
});

const generateXorY = (dim: string) => [
  injectFieldPred(dim, buildTypeCoercion(dim)),
  buildScaleOption(dim),
  injectFieldPred(dim, buildSpatialOptions(dim, spatialAggs)),
  injectNofieldPred(dim, buildSpatialOptions(dim, justCountAgg)),
];

// the configuration options consumed by the encoding panel
export const configurationOptions: any = {
  x: generateXorY('x'),
  y: generateXorY('y'),
  'row-meta': [],
  'column-meta': [],
  'repeat-meta': [],

  // repeaters
  facet: [injectFieldPred('facet', buildTypeCoercion('facet'))],
  row: [injectFieldPred('row', buildTypeCoercion('row'))],
  column: [injectFieldPred('column', buildTypeCoercion('column'))],
  // marks
  size: [
    injectFieldPred('size', buildTypeCoercion('size')),
    injectFieldPred('size', buildSpatialOptions('size', binningOptions)),
    injectNofieldPred('size', buildSpatialOptions('size', justCountAgg)),
  ],
  color: [
    injectFieldPred('color', buildTypeCoercion('color')),
    injectFieldPred('color', buildSpatialOptions('color', binningOptions)),
    injectNofieldPred('color', buildSpatialOptions('color', justCountAgg)),
  ],
  shape: [
    injectFieldPred('shape', buildTypeCoercion('shape')),
    injectFieldPred('shape', buildSpatialOptions('shape', binningOptions)),
  ],
  detail: [
    injectFieldPred('detail', buildTypeCoercion('detail')),
    injectFieldPred('detail', buildSpatialOptions('detail', binningOptions)),
  ],
  text: [
    injectFieldPred('text', buildTypeCoercion('text')),
    injectFieldPred('text', buildSpatialOptions('text', binningOptions)),
    injectNofieldPred('text', buildSpatialOptions('text', justCountAgg)),
  ],
};
