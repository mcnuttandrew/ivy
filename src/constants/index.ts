export const test = 'test';
type ON = 'on';
export const EDITOR_OPTIONS = {
  selectOnLineNumbers: true,
  automaticLayout: true,
  wordWrap: 'on' as ON,
  minimap: {
    enabled: false,
  },
  fontSize: 10,
  lineDecorationsWidth: 0,
  lineNumbersMinChars: 3,
};

export const TEXT_TYPE: {[x: string]: string} = {
  MEASURE: '123',
  DIMENSION: 'ABC',
  TIME: 'ðŸ•‘',
};

export const GRAMMAR_NAME = 'grammer';
export const GRAMMAR_DESC = 'Tableau-style grammar of graphics';

// export const
export const WIDGET_VALUES = 'Settings';
export const WIDGET_CONFIGURATION = 'Params';
export const JSON_OUTPUT = 'Output';
export const TEMPLATE_BODY = 'Body';

/* eslint-disable no-undef*/
export const PREVENT_ACCIDENTAL_LEAVE = process.env.NODE_ENV === 'production';
/* eslint-enable no-undef*/
