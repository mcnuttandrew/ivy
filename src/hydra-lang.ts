import {
  Template,
  TemplateMap,
  TemplateWidget,
  WidgetValidationQuery,
  DataTargetWidget,
  WidgetSubType,
  ListWidget,
  Shortcut,
  SwitchWidget,
  SliderWidget,
} from './templates/types';
import {trim} from './utils';

/**
 * Evaluate a hydra query, used for the widget validations and conditional checks
 * // TODO: this system doesn't support data type checking?
 * @param query - for now uses the special widget validation langugage
 * @param templateMap - the specification/variable values defined by the gui
 */
function evaluateQuery(query: WidgetValidationQuery, templateMap: TemplateMap): boolean {
  // TODO add a type check function to this
  const generatedContent = new Function('parameters', `return ${query}`);
  return Boolean(generatedContent(templateMap));
}

export function evaluateShortcut(shortcut: Shortcut, templateMap: TemplateMap): TemplateMap {
  const generatedContent = new Function('parameters', `return ${shortcut.shortcutFunction}`);
  const newMap = generatedContent(templateMap);
  return newMap;
}

interface ConditionalArgs {
  query: string;
  true: any;
  false: any;
  deleteKeyOnFalse?: boolean;
  deleteKeyOnTrue?: boolean;
}
function shouldUpdateContainerWithValue(
  queryResult: 'true' | 'false',
  conditional: ConditionalArgs,
): boolean {
  // if a conditional doesn't want that value to get added to the traversing object then ingnore it
  return (
    (queryResult === 'false' && conditional.deleteKeyOnFalse) ||
    (queryResult === 'true' && conditional.deleteKeyOnTrue)
  );
}

// syntax example
// {...example,
// mark: {
//   type: 'point',
//   tooltip: true,
//   color: {CONDITIONAL: {true: '[Single Color]', false: null, query: {Color: null}}},
// }}
/**
 * Walk across the tree and apply conditionals are appropriate,
 * example conditional syntax: {CONDITIONAL: {true: '[Single Color]', false: null, query: '!parameters.Color}}
 *
 * @param templateMap - the specification/variable values defined by the gui
 */
export function applyConditionals(templateMap: TemplateMap): any {
  return function walker(spec: any): any {
    // if it's array interate across it
    if (Array.isArray(spec)) {
      return spec.reduce((acc, child) => {
        if (child && typeof child === 'object' && child.CONDITIONAL) {
          const queryResult = evaluateQuery(child.CONDITIONAL.query, templateMap) ? 'true' : 'false';
          if (!shouldUpdateContainerWithValue(queryResult, child.CONDITIONAL)) {
            return acc.concat(walker(child.CONDITIONAL[queryResult]));
          } else {
            return acc;
          }
        }
        return acc.concat(walker(child));
      }, []);
    }
    // check if it's null or not an object return
    if (!(typeof spec === 'object' && spec !== null)) {
      return spec;
    }
    // if the object being consider is itself a conditional evaluate it
    if (typeof spec === 'object' && spec.CONDITIONAL) {
      const queryResult = evaluateQuery(spec.CONDITIONAL.query, templateMap) ? 'true' : 'false';
      if (!shouldUpdateContainerWithValue(queryResult, spec.CONDITIONAL)) {
        return walker(spec.CONDITIONAL[queryResult]);
      } else {
        return null;
      }
    }
    // otherwise looks through its children
    return Object.entries(spec).reduce((acc: any, [key, value]: any) => {
      // if it's a conditional, if so execute the conditional
      if (value && typeof value === 'object' && value.CONDITIONAL) {
        const queryResult = evaluateQuery(value.CONDITIONAL.query, templateMap) ? 'true' : 'false';
        if (!shouldUpdateContainerWithValue(queryResult, value.CONDITIONAL)) {
          acc[key] = walker(value.CONDITIONAL[queryResult]);
        }
      } else {
        acc[key] = walker(value);
      }
      return acc;
    }, {});
  };
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function applyQueries(template: Template, templateMap: TemplateMap): TemplateWidget<any>[] {
  const widgetMap = template.widgets.reduce((acc: any, widget) => {
    acc[widget.widgetName] = true;
    return acc;
  }, {});

  const validWidgetNames = template.widgetValidations.reduce((acc: {[x: string]: boolean}, validation) => {
    const queryResult = evaluateQuery(validation.query, templateMap);
    acc[validation.queryTarget] = validation.queryResult === 'show' ? queryResult : !queryResult;
    return acc;
  }, widgetMap);
  return template.widgets.filter(widget => validWidgetNames[widget.widgetName]);
}

/**
 * Apply values from templatemap (specification) to template
 * Important to do it this way and not via json parse because values might be parts of strings
 * Example {"field": "datum.[VARIABLENAME]"}
 * @param code
 * @param templateMap - the specification/variable values defined by the gui
 */
export const setTemplateValues = (code: string, templateMap: TemplateMap): string => {
  const filledInSpec = Object.entries(templateMap).reduce((acc: string, keyValue: any) => {
    const [key, value] = keyValue;
    if (trim(value) !== value) {
      // this supports the weird HACK required to make the interpolateion system
      // not make everything a string
      return acc
        .replace(new RegExp(`"\\[${key}\\]"`, 'g'), value || 'null')
        .replace(new RegExp(`\\[${key}\\]`, 'g'), trim(value) || 'null');
    }
    const reg = new RegExp(`"\\[${key}\\]"`, 'g');
    return acc.replace(reg, (Array.isArray(value) && JSON.stringify(value)) || value || 'null');
  }, code);
  return filledInSpec;
};

/**
 * Generate a list of the missing fields on a template
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function getMissingFields(template: Template, templateMap: TemplateMap): string[] {
  const requiredFields = template.widgets
    .filter(d => d.widgetType === 'DataTarget' && (d as TemplateWidget<DataTargetWidget>).widget.required)
    .map(d => d.widgetName);
  const missingFileds = requiredFields
    .map((fieldName: string) => ({fieldName, value: !templateMap[fieldName]}))
    .filter((d: any) => d.value)
    .map(d => d.fieldName);

  return missingFileds;
}

/**
 * Figure out if the all of the required fields have been filled in
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function checkIfMapComplete(template: Template, templateMap: TemplateMap): boolean {
  const missing = getMissingFields(template, templateMap);
  return missing.length === 0;
}

/**
 * for template map holes that are NOT data columns, fill em as best you can
 * @param template
 */
export function constructDefaultTemplateMap(template: Template): TemplateMap {
  return template.widgets.reduce((acc: any, w: TemplateWidget<WidgetSubType>) => {
    let value = null;
    if (w.widgetType === 'MultiDataTarget') {
      value = [];
    }
    if (w.widgetType === 'Text') {
      return acc;
    }
    if (w.widgetType === 'List') {
      value = (w as TemplateWidget<ListWidget>).widget.defaultValue;
    }
    if (w.widgetType === 'Switch') {
      const localW = w as TemplateWidget<SwitchWidget>;
      value = localW.widget.defaultsToActive ? localW.widget.activeValue : localW.widget.inactiveValue;
    }
    if (w.widgetType === 'Slider') {
      value = (w as TemplateWidget<SliderWidget>).widget.defaultValue;
    }
    acc[w.widgetName] = value;
    return acc;
  }, {});
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function evaluateHydraProgram(template: Template, templateMap: TemplateMap): any {
  // 1. apply variables to string representation of code
  const interpolatedVals = setTemplateValues(template.code, templateMap);
  // 2. parse to json
  // const parsedJson = JSON.parse(interpolatedVals);
  let parsedJson = null;
  try {
    parsedJson = JSON.parse(interpolatedVals);
  } catch (e) {
    console.log(e, 'crash');
    parsedJson = {};
  }
  // 3. evaluate inline conditionals
  const evaluatableSpec = applyConditionals(templateMap)(parsedJson);
  // 4. return
  return evaluatableSpec;
}
