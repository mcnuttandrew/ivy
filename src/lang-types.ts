/**
 * Type necessary to define the validator pull over here
 * in order to prevent stack overflow on the schema builder
 */
export type DataType = 'MEASURE' | 'DIMENSION' | 'TIME' | 'CUSTOM';

export type WidgetType =
  | 'DataTarget'
  | 'MultiDataTarget'
  | 'List'
  | 'Switch'
  | 'Text'
  | 'Slider'
  | 'Section'
  | 'Shortcut'
  | 'FreeText';
export interface Widget<T> {
  /**
   *   The name of widget to be used, this name will be swapped into the code string, must be unqiue
   */
  name: string;
  /**
   * The name to be shown in the GUI, does not have to be unique.
   */
  displayName?: string;
  /**
   * The type of the widget to be used, this defined the specific gui item that the user interacts with
   */
  type: WidgetType;
  config: T;
  /**
   * Sometimes you want to decative certain values depending on the state of the UI
   * This advanced features allows you to do that
   */
  conditions?: Condition[];
}
export interface DataTargetWidget {
  allowedTypes: DataType[];
  required: boolean;
}
export interface MultiDataTargetWidget {
  allowedTypes: DataType[];
  minNumberOfTargets?: number;
  maxNumberOfTargets?: number;
  required: boolean;
}
export interface ListWidget {
  allowedValues: ({display: string; value: string} | string)[];
  defaultValue?: string;
}
export interface SwitchWidget {
  active: string;
  inactive: string;
  defaultsToActive: boolean;
}
export interface TextWidget {
  text: string;
}
export type SectionWidget = null;
export interface SliderWidget {
  minVal: number;
  maxVal: number;
  step?: number;
  defaultValue: number;
}

export interface Shortcut {
  label: string;
  shortcutFunction: string;
}
export interface ShortcutsWidget {
  shortcuts: Shortcut[];
}
export type FreeTextWidget = {
  useParagraph: boolean;
};
export type WidgetSubType =
  | DataTargetWidget
  | MultiDataTargetWidget
  | ListWidget
  | SwitchWidget
  | TextWidget
  | SliderWidget
  | SectionWidget
  | ShortcutsWidget
  | FreeTextWidget;

/**
 * The main configuration object for templates
 */
export interface Template {
  /**
   * The language of the code, determines how the code will be interpreted.
   * Ivy currently supports vega, vega-lite, and it's own data table system
   *
   *  __Default value:__ `vega-lite`
   */
  templateLanguage: string;

  /**
   * The name of the template. Template names must be unique, so this can over-ride other extant templates
   */
  templateName: string;

  /**
   * The description that users will see in the chart chooser gallery
   */
  templateDescription?: string;

  /**
   * The creator of the template
   */
  templateAuthor: string;

  /**
   * The code to be interpreted by the renderer
   */
  code: string;

  /**
   * The mechanism by which users interact with your template
   */
  widgets: Widget<WidgetSubType>[];

  /**
   * Advanced tool for providing special extra cards
   */
  customCards?: CustomCard[];

  /**
   * Whether or not to allow view materialization / fan out
   */
  disallowFanOut?: boolean;
}

/**
 * Convience container type for the general template widget case
 */
export type GenWidget = Widget<WidgetSubType>;

export type CustomCard = {name: string; description: string};
/**
 * A widget Condition query, executed raw javascript. Parameter values (the value of the current ui)
 * is accessed through parameters.VALUE. E.g. if you wanted to construct a predicate that check if
 * there wasn't a current value for the x dimension called xDim you could do "!parameters.xDim"
 */
export type ConditionQuery = string;

/**
 * What to do in response to the result of the query, should be either 'hide' or 'show'
 */
export type QueryResult = 'show' | 'hide';
export interface Condition {
  /**
   * What to do in response to the result of the query, should be either 'hide' or 'show'
   */
  queryResult: QueryResult;

  /**
   * A widget Condition query, executed raw javascript. Parameter values (the value of the current ui)
   * is accessed through parameters.VALUE. E.g. if you wanted to construct a predicate that check if
   * there wasn't a current value for the x dimension called xDim you could do "!parameters.xDim"
   */
  query: string;
}
