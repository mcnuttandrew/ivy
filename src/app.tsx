import React from 'react';
import ReactDOM from 'react-dom';
import AppWrap from './components/app-wrap';
import setupMonaco from './utils/monaco';
import {PREVENT_ACCIDENTAL_LEAVE} from './constants/CONFIG';
import {retrieveThumbnails} from './thumbnail';

import './stylesheets/main.css';
import './stylesheets/rc-slider.css';
import 'rc-tooltip/assets/bootstrap_white.css';

setupMonaco();
retrieveThumbnails();

ReactDOM.render(<AppWrap />, document.querySelector('#root-container'));
if (document.querySelector('.loading-msg')) {
  document.querySelector('.loading-msg').remove();
}

if (PREVENT_ACCIDENTAL_LEAVE) {
  window.onbeforeunload = (): string => 'Are you sure you want to leave?';
}
