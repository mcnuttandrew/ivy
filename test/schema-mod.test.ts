import {modifyJSONSchema} from '../src/hydra-lang';

import Ajv from 'ajv';
import polestarTemplate from '../src/templates/example-templates/polestar-template';
/* eslint-disable @typescript-eslint/no-var-requires */
import {writeFile} from 'hoopoe';
const draft6Schema = require('ajv/lib/refs/json-schema-draft-06.json');
const vegaLiteSchema = require('vega-lite/build/vega-lite-schema.json');
const unitVisSchema = require('unit-vis/unit-vis-schema.json');
/* eslint-enable @typescript-eslint/no-var-requires */
// const valid = ajv.validate(vegaLiteSchema, polestarTemplate.code);
const ajv = new Ajv({
  allErrors: true,
  verbose: false,
  format: 'uri-reference',
  extendRefs: 'fail',
  schemaId: 'auto', // for draft 04 and 06 schemas
  validateSchema: true,
});
ajv.addMetaSchema(draft6Schema);
ajv.addFormat('color-hex', () => true);

const COMMON = {
  $schema: 'https://vega.github.io/schema/vega-lite/v4.json',
  description: 'A simple bar chart with embedded data.',
  data: {
    values: [
      {a: 'A', b: 28},
      {a: 'B', b: 55},
      {a: 'C', b: 43},
      {a: 'D', b: 91},
      {a: 'E', b: 81},
      {a: 'F', b: 53},
      {a: 'G', b: 19},
      {a: 'H', b: 87},
      {a: 'I', b: 52},
    ],
  },
};

const EXAMPLE_VL_SCHEMA = {
  ...COMMON,
  fuck: 'this',
  mark: 'bar',
  encoding: {
    x: {field: 'a', type: 'ordinal'},
    y: {field: 'b', type: 'quantitative'},
  },
};

const EXAMPLE_VL_SCHEMA_MODIFIED_CONDITIONAL: any = {
  ...COMMON,
  mark: 'bar',
  encoding: {
    x: {
      field: {
        CONDITIONAL: {
          query: 'parameters.x',
          true: '[X]',
          //   false: null,
        },
      },
      type: 'quantitative',
    },
    y: {field: 'b', type: 'quantitative'},
  },
};

const EXAMPLE_VL_SCHEMA_MODIFIED_INTERPOLANT: any = {
  ...COMMON,
  mark: 'bar',
  encoding: {
    x: {field: 'a', type: '[xType]'},
    y: {field: 'b', type: '[yType]'},
  },
};

const UNIT_VIS_EXAMPLE = {
  $schema: 'https://unit-vis.netlify.com/assets/unit-vis-schema.json',
  fuck: 'this',
  data: {url: 'data/titanic3.csv'},
  layouts: [{subgroup: {type: 'flatten'}}],
  mark: {color: {key: 'survived_text', type: 'categorical'}},
};

test('#modifyJSONSchema', async function() {
  //   const uvSchema = modifyJSONSchema(unitVisSchema);
  //   console.log(JSNuvSchema);
  // catches unused invalid top level key
  //   const validity1 = ajv.validate(unitVisSchema, UNIT_VIS_EXAMPLE);
  //   expect(validity1).toBe(true);
  //   expect(ajv.errors).toMatchSnapshot();

  const vlSchema = modifyJSONSchema(vegaLiteSchema);
  console.log(JSON.stringify(vlSchema, null, 2));
  // catches unused invalid top level key
  const validity1 = ajv.validate(vlSchema, EXAMPLE_VL_SCHEMA);
  //   console.log(ajv.errors);
  expect(validity1).toBe(false);
  expect(ajv.errors).toMatchSnapshot();

  // doesnt invalidate interpolants
  const validity2 = ajv.validate(vlSchema, EXAMPLE_VL_SCHEMA_MODIFIED_INTERPOLANT);
  expect(validity2).toBe(true);
  expect(ajv.errors).toBe(null);

  // doesnt invalidate conditionals
  const validity3 = ajv.validate(vlSchema, EXAMPLE_VL_SCHEMA_MODIFIED_CONDITIONAL);
  expect(validity3).toBe(true);
  expect(ajv.errors).toBe(null);
});
