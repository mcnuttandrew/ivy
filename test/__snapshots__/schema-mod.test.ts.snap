// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#modifyJSONSchema 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "encoding",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'facet'",
    "params": Object {
      "missingProperty": "facet",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'spec'",
    "params": Object {
      "missingProperty": "spec",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'layer'",
    "params": Object {
      "missingProperty": "layer",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "encoding",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'repeat'",
    "params": Object {
      "missingProperty": "repeat",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'spec'",
    "params": Object {
      "missingProperty": "spec",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "encoding",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'concat'",
    "params": Object {
      "missingProperty": "concat",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "encoding",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'vconcat'",
    "params": Object {
      "missingProperty": "vconcat",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "fuck",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "mark",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "encoding",
    },
    "schemaPath": "#/additionalProperties",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'hconcat'",
    "params": Object {
      "missingProperty": "hconcat",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "anyOf",
    "message": "should match some schema in anyOf",
    "params": Object {},
    "schemaPath": "#/anyOf",
  },
]
`;

exports[`#modifyJSONSchema 2`] = `null`;

exports[`#modifyJSONSchema 3`] = `null`;
